runtime/lua/vim/fs.lua

```diff
@@ -349,22 +349,24 @@ function M.normalize(path, opts)
   })
 
   if path:sub(1, 1) == '~' then
-    local home = vim.uv.os_homedir() or '~'
-    if home:sub(-1) == '\\' or home:sub(-1) == '/' then
-      home = home:sub(1, -2)
-    end
-    path = home .. path:sub(2)
+    path = (vim.uv.os_homedir() or '~') .. '/' .. path:sub(2)
   end
 
   if opts.expand_env == nil or opts.expand_env then
     path = path:gsub('%$([%w_]+)', vim.uv.os_getenv)
   end
 
-  path = path:gsub('\\', '/'):gsub('/+', '/')
-  if iswin and path:match('^%w:/$') then
-    return path
+  local first = path:sub(1, 1)
+  path = (first == '\\' and '/' or first) .. path:sub(2):gsub('[\\/]+', '/'):gsub('/$', '')
```

Comment:
Shouldn't the `\` -> `/` be unconditional?

```suggestion
  path = path:gsub('\\', '/')
  path = path:sub(1, 1) .. path:sub(2):gsub('/+', '/')
```

Reply:
This is wrong. We must have `:gsub('/$', '')` to strip the trailing slash (eg `foo/bar/`)

Your solution will then require to call `gsub` 3 times, which is not efficient.

Iiuc, we should assume this function will be called _a lot_ of times, so we should take performance over code readability, and tbh, I think my solution is quite clean ngl.

Reply:
If you want to avoid another gsub then I suggest not doing `:gsub('/$', '')` and instead do 

```lua
if path:sub(-1) == '/' then
  path = path:sub(1, -2)
end
```

> so we should take performance over code readability, and tbh, I think my solution is quite clean ngl.

Nothing indicates this function is a bottleneck in any code path. And if you're dealing with the file system any `gsub`'s are going to be completely dwarfed by calls out to the system.

Reply:
I don't think this is clean and I think it makes the code less readable. So I see no reason to change my code.

> Nothing indicates this function is a bottleneck in any code path. And if you're dealing with the file system any `gsub`'s are going to be completely dwarfed by calls out to the system.

I see.

Reply:
That's unfortunate. Personally, I don't think squashing many terms into a single line that do multiple things is clean at all. It's short, but it isn't clean or easy to understand. By breaking it down into multiple independent operations, the reader can comprehend each distinct step separately.

And if we keep the additional `gsub`, because realistically this isn't going meaningfully affect performance, then I fail to see how:

```lua
local first = path:sub(1, 1)
path = (first == '\\' and '/' or first) .. path:sub(2):gsub('[\\/]+', '/'):gsub('/$', '')
```

Is cleaner than:
```lua
path = path:gsub('\\', '/')  -- Normalize slashes
path = path:sub(1, 1) .. path:sub(2):gsub('/+', '/'):gsub('/$', '')  -- Remove redundant slashes
```

Reply:
I don't think your code explains it well enough because _Remove redundant slashes_ does not explain `sub(1, 1)` and its quite unclear anyways.

It doesn't document that `//` will result in `/` but `//foo` will be `//foo`, and `///foo` will be `//foo`.

I didn't want to say that my code is cleaner, but I meant any code will not be clear anyways, and didn't find a reason to introduce another gsub for little benifit.